sherpa/data.py

```diff
@@ -1048,25 +1106,16 @@ def indep(self):
 
         Returns
         -------
-        tuple of array_like
+        tuple of array_like or None
 
         """
         return self._data_space.get().grid
 
-    def _clear_filter(self):
-        """Clear out the existing filter.
-
-        This is designed for use by @indep.setter.
-
-        """
-
-        # This is currently a no-op. It may beover-ridden by a
-        # subclass.
-        #
-        pass
-
     @indep.setter
-    def indep(self, val):
+    def indep(self,
+              val: Union[tuple[ArrayType, ...],
+                         tuple[None, ...]]
+              ) -> None:
 
```

Comment:
I know this is beyond the scope if the current PR, but just for my curiosity: Wha'ts the use case for setting indep to None? If I have no independent axis, I don't have any data?

Reply:
So, I am basically trying to type the existing code, and it allows for this. I personally would like to remove the "can be `None`" ability for the independent and dependent axes completely, but this is used in the existing fake-PHA data code, and there are traces throughout the code for allowing `None` values to be returned. So, follow-up PRs could be done to simplify this - for instance, at the moment you can do

```
>>> from sherpa.data import Data1D
>>> d = Data1D('x', None, None)
>>> d.dep = None
>>> d.indep = (None, )
>>> d.dep = [1, 2, 3]
>>> d.dep = None
>>> d.dep = [1, 2, 3, 4]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dburke/sherpa/sherpa-xspec/sherpa/utils/__init__.py", line 137, in __setattr__
    object.__setattr__(self, name, val)
  File "/home/dburke/sherpa/sherpa-xspec/sherpa/data.py", line 1030, in dep
    self.y = val
    ^^^^^^
  File "/home/dburke/sherpa/sherpa-xspec/sherpa/utils/__init__.py", line 137, in __setattr__
    object.__setattr__(self, name, val)
  File "/home/dburke/sherpa/sherpa-xspec/sherpa/data.py", line 997, in y
    self._set_related("y", val, check_mask=False)
  File "/home/dburke/sherpa/sherpa-xspec/sherpa/data.py", line 983, in _set_related
    raise DataErr('mismatchn', 'independent axis', attr, nelem, nval)
sherpa.utils.err.DataErr: size mismatch between independent axis and y: 3 vs 4
>>> d.indep = ([45, 99, 203], )
```

and I could see us removing the ability to set either `indep` or `dep` to `None` (i.e. only allowing the `DatanD(name, None, ..., None)` call). However, there's always the compatibility concern, but that's for future-PR writer to worry about!

Reply:
Thanks for the explanation.